File: lecture3/node_demo_pkg/scripts/node_demo.py
#!/usr/bin/env python3

"""
ROS 2 Node Entry Point.

This script initializes and runs a ROS 2 node using the `NodeDemoInterface` class.
It sets up the ROS 2 environment, creates a subscriber or publisher node, and 
properly shuts down ROS 2 before exiting.

Author: Zeid Kootbally
"""

import rclpy
from node_demo_pkg.node_demo_interface import NodeDemoInterface


def main(args=None):
    """Initialize and run the ROS 2 node.

    This function initializes the ROS 2 system, creates an instance of 
    `NodeDemoInterface`, and shuts down ROS 2 after execution. 

    Note:
        - `rclpy.spin(node)` is currently commented out, so the node does 
          not actively process callbacks.
        - Uncomment `rclpy.spin(node)` to keep the node running.
    
    Args:
        args (list, optional): Command-line arguments for ROS 2 initialization. 
                               Defaults to None.
    """
    rclpy.init(args=args)
    node = NodeDemoInterface("node_demo_py")
    rclpy.shutdown()

    # Uncomment the following block to enable node execution:
    # try:
    #     rclpy.spin(node)  # Keeps the node running
    # except Exception as e:
    #     rclpy.logging.get_logger("node_demo_py").error(f"Error initializing node: {e}")
    # except KeyboardInterrupt:
    #     # Log a message when the node is manually terminated
    #     node.get_logger().warn("Keyboard interrupt detected")
    # finally:
    #     node.get_logger().warn("Exited spin()")
    #     rclpy.shutdown()

if __name__ == "__main__":
    main()

File: lecture3/node_demo_pkg/src/node_demo.cpp
/**
 * @file main.cpp
 * @brief Entry point for the ROS 2 node using NodeDemoInterface.
 *
 * This file initializes the ROS 2 system, creates a shared instance of the 
 * `NodeDemoInterface` node, and manages the ROS 2 lifecycle.
 */

#include <rclcpp/rclcpp.hpp>
#include "node_demo_pkg/node_demo_interface.hpp"

/**
 * @brief Main function for the ROS 2 node.
 *
 * This function initializes ROS 2, creates an instance of `NodeDemoInterface`, 
 * and prepares the node for execution. The `rclcpp::spin(node);` line is commented 
 * out, meaning the node is currently not being actively spun.
 *
 * @param argc Number of command-line arguments.
 * @param argv Array of command-line arguments.
 * @return int Returns 0 upon successful execution.
 */
int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);

    auto node = std::make_shared<NodeDemoInterface>("node_demo_cpp");
    // rclcpp::spin(node);  ///< Uncomment to keep the node running.

    rclcpp::shutdown();
    return 0;
}
